Sources/Utility/ProcessSet.swift

```diff
@@ -0,0 +1,127 @@
+/*
+ This source file is part of the Swift.org open source project
+ 
+ Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+ 
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Dispatch
+import Foundation
+import Basic
+
+typealias Thread = Basic.Thread
+
+public enum ProcessSetError: Swift.Error {
+    /// The process group was cancelled and doesn't allow adding more processes.
+    case cancelled
+}
+
+/// A process set is a small warpper for collection of processes.
+/// 
+/// This class is thread safe.
+public final class ProcessSet {
+
+    /// Array to hold the processes.
+    private var processes: Set<Process> = []
+
+    /// Queue to mutate internal states of the process group.
+    private let serialQueue = DispatchQueue(label: "org.swift.swiftpm.process-set")
+
+    /// If the process group was asked to cancel all active processes.
+    private var cancelled = false
+
+    /// The timeout (in seconds) after which the processes should be killed if they don't respond to SIGINT.
+    public let killTimeout: Double
+
+    /// Condition to block kill thread until timout.
+    private var killingCondition = Condition()
+
+    /// Boolean predicate for killing condition.
+    private var shouldKill = false
+
+    /// Create a process set.
+    public init(killTimeout: Double = 5) {
+        self.killTimeout = killTimeout
+    }
+
+    /// Add a process to the process set. This method will throw if the process set is already terminated.
```

Comment:
Yes, I meant to say if someone called `terminate()`, this method will start throwing